docs/design/features/StringDeduplication.md

```diff
@@ -0,0 +1,196 @@
+
+An example of a user implemented string deduplication is Roslyn’s [StringTable utility](https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/InternalUtilities/StringTable.cs).
+
+#### **Customer impact estimation and validation**
+
+As a general rule we want to have this for all features we add to the runtime. 
+
+Issue #[9022](https://github.com/dotnet/runtime/issues/9022) It mentioned some general data: 
+
+“The expectation is that typical apps have 20% of their GC heap be strings. Some measurements we have seen is that for at least some applications, 10-30% of strings all may be duplicated, so this might save 2-3% of the GC heap. Not huge, but the feature is not that difficult either.”
+
+But keep in mind that this is an *average* - specific scenarios could have much higher percentage of duplicated strings. I have heard from customers who had much more dramatic heap reduction when they eliminated duplicated strings. Those are scenarios that would benefit from this feature the most.
+
+There are 2 sources of data we could get –
+
+- Our dump database. We can ask the TI team to help us look at the general percent of duplicated strings on the GC heap. It would be useful to also provide tooling to help users to check how much space duplicated strings take. Some internal folks have written tooling for this but it requires work to ship.
+- We already know of teams that can benefit from deduping. We should start a conversation with them to collect some data to have a rough idea about the baseline and make sure we have a way to verify private builds with them.
+
+#### **Design outline**
+
+This is an opt in feature. When the runtime detects it’s turned on, it creates a dedup thread to do the work. 
+
+Detection of duplicated strings is done by looking into a hash table. The key into this hash table is the hash code of the content of a string. Detailed description of this detection is later in this doc. 
+
+As the dedup thread goes through the old generations linearly, it looks for references to a string object (denoted by the method table) and either calculates or looks up the hash code of that string to see if it already exists in the hash table. If so it will attempt to change the reference to point to that string with a CAS operation. If this fails, which means some other thread changed the reference at the mean time, we simply ignore this and move on. We expect the CAS failure rate to be very low.
+
+Since the new string reference we will write to the heap has the exact same type info, it means BGC can read either the old or the new reference and will get the same size info so deduping would not conflict with BGC concurrent mark. It does not conflict with BGC sweep because BGC sweep does not read references inside an object.
```

Comment:
I am not sure what this is trying to say.

Is this just trying to say that the string deduping runs independently on BCG, and thus there are no concerns about conflicts between the two?

Reply:
right, this just gives the reason why it doesn't cause problems with the BGC which also goes through the heap. it's important from GC's POV but from other components' POV this is just a detail they can choose to not care about.